ntp-proto/src/algorithm/kalman/mod.rs

```diff
@@ -0,0 +1,358 @@
+            // jump
+            self.clock
+                .step_clock(NtpDuration::from_seconds(change))
+                .unwrap();
+            for (state, _) in self.peers.values_mut() {
+                state.process_offset_steering(change)
+            }
+            info!("Jumped offset by {}ms", change * 1e3);
+            None
+        } else {
+            // start slew
+            let freq = self
+                .algo_config
+                .slew_max_frequency_offset
+                .min(change.abs() / self.algo_config.slew_min_duration);
+            let duration = NtpDuration::from_seconds(change.abs() / freq);
+            info!(
+                "Slewing by {}ms over {}s",
+                change * 1e3,
+                duration.to_seconds()
+            );
+            Some(self.change_desired_frequency(-freq * change.signum(), freq_delta) + duration)
+        }
+    }
+
+    fn change_desired_frequency(&mut self, new_freq: f64, freq_delta: f64) -> NtpTimestamp {
+        let change = self.desired_freq - new_freq + freq_delta;
+        self.desired_freq = new_freq;
+        self.steer_frequency(change)
+    }
+
+    fn steer_frequency(&mut self, change: f64) -> NtpTimestamp {
+        self.freq_offset = (1.0 + self.freq_offset) * (1.0 + change) - 1.0;
+        self.clock.set_frequency(self.freq_offset).unwrap();
+        let freq_update = self.clock.now().unwrap();
+        for (state, _) in self.peers.values_mut() {
+            state.process_frequency_steering(freq_update, change)
+        }
+        info!(
+            "Changed frequency, current steer {}ppm, desired freq {}ppm",
+            self.freq_offset * 1e6,
+            self.desired_freq * 1e6,
+        );
+        freq_update
+    }
+
+    fn update_desired_poll(&mut self) {
+        self.timedata.poll_interval = self
+            .peers
+            .values()
+            .map(|(state, _)| state.get_desired_poll(&self.config.poll_limits))
+            .min()
+            .unwrap_or(self.config.poll_limits.max);
+    }
+}
+
+impl<C: NtpClock, PeerID: Hash + Eq + Copy + Debug> TimeSyncController<C, PeerID>
+    for KalmanClockController<C, PeerID>
+{
+    type AlgorithmConfig = AlgorithmConfig;
+
+    fn new(clock: C, config: SystemConfig, algo_config: Self::AlgorithmConfig) -> Self {
+        // Setup clock
+        clock
+            .disable_ntp_algorithm()
+            .expect("Unable to change system time");
+        clock
+            .status_update(NtpLeapIndicator::Unknown)
+            .expect("Unable to update clock");
+        clock
+            .set_frequency(0.0)
+            .expect("Unable to set system clock frequency");
+
+        KalmanClockController {
+            peers: HashMap::new(),
+            clock,
+            config,
+            algo_config,
+            freq_offset: 0.0,
+            desired_freq: 0.0,
+            timedata: TimeSnapshot::default(),
+            in_startup: false,
```

Comment:
should this not be `true`? based on my searching, `in_startup` is never true currently.

Reply:
Good find, I'll fix it and add tests for this flag and its functioning.

Reply:
Fixed